File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        long long  mod = 1e9+7;
        int n  = arr.size();
        
        vector<long long> left(n,0),right(n,0);

        stack<pair<int,int>> sRight,sLeft;

        for(int i =0;i<arr.size();i++)
        {
            int count = 1;
            while(!sLeft.empty() && sLeft.top().first>=arr[i])
            {
                count+=sLeft.top().second ; //Adding the count of the first element also
                sLeft.pop();
            }
            sLeft.push({arr[i],count});
            left[i] = count;
        }
        
        for(int i =n-1;i>=0;i--)
        {
            int count = 1;
            while(!sRight.empty() && sRight.top().first>arr[i])
            {
                count+=sRight.top().second ; //Adding the count of the first element also
                sRight.pop();
            }
            sRight.push({arr[i],count});
            right[i] = count;
        }
        
        long long int res = 0 ;
        for(int i=0;i<arr.size();i++)
        {
            
            res += (arr[i]*left[i]*right[i]);
            res = res%mod;
            
        }
        return res;
    }
};

File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        #brute for would be find all subarrays and then find minimums in each and add them

        #Optimised solution

        def nextSmallestElement(arr):
            n = len(arr)
            nse = [0]*n
            stack = []

            for i in range(n-1,-1,-1):
                if not stack:
                    nse[i]=n
                
                elif stack and arr[stack[-1]] < arr[i]:
                    nse[i] = stack[-1]
                
                elif stack and arr[stack[-1]] >= arr[i]:
                    while stack and arr[stack[-1]] >= arr[i]:
                        stack.pop()
                    
                    if not stack:
                        nse[i]=n
                    else:
                        nse[i] = stack[-1]
                
                stack.append(i)
        
            return nse

        def prevSmallestElement(arr):
            n = len(arr)
            pse = [0]*n
            stack = []

            for i in range(0,n):
                if not stack:
                    pse[i]=-1
                
                elif stack and arr[stack[-1]] <= arr[i]:
                    pse[i] = stack[-1]
                
                elif stack and arr[stack[-1]] > arr[i]:
                    while stack and arr[stack[-1]] > arr[i]:
                        stack.pop()
                    
                    if not stack:
                        pse[i]=-1
                    else:
                        pse[i] = stack[-1]
                
                stack.append(i)
        
            return pse
        pse= prevSmallestElement(arr)
        nse= nextSmallestElement(arr)

        print(nse,pse)

        sum = 0
        for i in range(len(arr)):
            left = i-pse[i]
            right = nse[i]-i
            sum = (sum+ (right*left*arr[i]))%(10**9+7)
        return sum